src/state.rs

```diff
@@ -41,7 +40,29 @@ use hole_punching::HolePunchServer;
 use util::ip_from_socketaddr;
 use util;
 
-pub type Closure = Box<FnBox(&mut State) + Send>;
+// Closure is a wapper around boxed closures that tries to work around the fact
+// that it is not possible to call Box<FnOnce> in the current stable rust.
+// The idea is to wrap the FnOnce in something that implements FnMut using some
+// dirty tricks, because Box<FnMut> is fine to call.
+//
+// This workaround can be removed once FnBox becomes stable or Box<FnOnce>
+// becomes usable.
+pub struct Closure(Box<FnMut(&mut State) + Send>);
+
+impl Closure {
+    pub fn new<F : FnOnce(&mut State) + Send + 'static>(f: F) -> Closure {
+        let mut f = Some(f);
+        Closure(Box::new(move |state: &mut State| {
```

Comment:
Just finished reading the thread and went to read your code again. Clever solution.